File: code-examples/MATLAB/slam/isam/isam_incremental_update_example.m
clear all;
more on;

disp('random measurement matrix');
A = rand(10,4)

disp('QR factorization using Givens rotations');
[Q,R] = QR(A)

disp('note that R is indeed equal to the information matrix square root (up to sign flip of rows)')
R_chol = chol(A'*A, 'upper')

disp('here is the transformed measurement matrix');
At = Q'*A;
At(abs(At)<abs(2*eps)) = 0

disp('suppose now that we add a new measurement row');
At_aug = [At; 0 0 0 1]

disp('we can incrementally obtain the new factorization using one Givens rotation');
alpha = At_aug(4,4);
beta = At_aug(11,4);
if beta == 0
    disp('beta = 0');
    c = 1; s = 0;
elseif abs(beta) > abs(alpha)
    disp('|beta| > |alpha|');
    c = -alpha/beta/sqrt(1+(alpha/beta)^2);
    s = 1/sqrt(1+(alpha/beta)^2);
else
    disp('otherwise');
    c = 1/sqrt(1+(beta/alpha)^2);
    s = -beta/alpha/sqrt(1+(beta/alpha)^2);
end
G = eye(11);
G(4,4) = c;
G(4,11) = s;
G(11,4) = -s;
G(11,11) = c

disp('incrementally applying the Givens rotation to our augmented system')
gAt_aug = G'*At_aug

disp('compare this with the batch QR factorization of At_aug');
[Qt_aug,Rt_aug] = QR(At_aug)

disp('and also compare with the Cholesky factorization of the augmented system');
R_aug_chol = chol(At_aug'*At_aug, 'upper')

more off;

File: code-examples/Python/slam/isam_incremental_update_example.py
#!usr/bin/env python

import numpy as np
from QR import QR_function

if __name__ == "__main__":
    print('Random Measurement Matrix:')
    A = np.random.randn(10, 4)
    print(A)

    print('QR Factorization Using Givens Rotations:')
    Q, R = QR_function(A)
    print('Q\n', Q, '\nR\n', R)

    print('Note that R is indeed equal to the information matrix square root (up to sign flip of rows)')
    R_chol = np.linalg.cholesky(np.dot(A.T, A)).T
    print(R_chol)

    print('Here is the transformed measurement matrix:')
    At = np.dot(Q.T, A)
    At[abs(At) < abs(2*np.finfo(float).eps)] = 0
    print(At)

    print('Suppose now that we ass a new measurement row:')
    At_aug = np.vstack((At, np.array([0, 0, 0, 1])))
    print(At_aug)

    print('We can incrementally obtain the new factorization using one Givens rotation:')
    alpha = At_aug[3, 3]
    beta = At_aug[10, 3]
    if beta == 0:
        print('beta = 0')
        c = 1
        s = 0
    elif abs(beta) > abs(alpha):
        print('|beta| > |alpha|')
        c = -alpha / beta / np.sqrt(1 + (alpha / beta)**2)
        s = 1 / np.sqrt(1 + (alpha / beta)**2)
    else:
        print('Otherwise')
        c = 1 / np.sqrt(1 + (alpha / beta) ** 2)
        s = -beta / alpha / np.sqrt(1 + (alpha / beta) ** 2)
    G = np.eye(11)
    G[3, 3] = c
    G[3, 10] = s
    G[10, 3] = -s
    G[10, 10] = c
    print(G)

    print('Incrementally applying the Givens rotation to our augmented system:')
    gAt_aug = np.dot(G.T, At_aug)
    print(gAt_aug)

    print('Compare this with the batch QR factorization of At_aug:')
    Qt_aug, Rt_aug = np.linalg.qr(At_aug)
    print('Qt_aug\n', Qt_aug, '\nRt_aug\n', Rt_aug)

    print('And also compare with the Cholesky factorization of the augmented system:')
    R_aug_chol = np.linalg.cholesky(np.dot(At_aug.T, At_aug)).T
    print(R_aug_chol)

